File: frontend/lib/api/rules.ts
import { apiFetch } from './client';
import type {
  MerchantRuleCreate,
  MerchantRuleOut,
  MerchantRuleUpdate,
} from './types';

/** 后端返 {items: MerchantRuleOut[], total: number}。
 *  Task 13 inline 写在这里(types.ts 未导出),Task 20 完整 CRUD 时如需扩展再统一。 */
export interface MerchantRuleListOut {
  items: MerchantRuleOut[];
  total: number;
}

/** GET /api/rules — 列表(无分页 query;后端返全量,默认按 priority 排序)。 */
export function listRules(): Promise<MerchantRuleListOut> {
  return apiFetch<MerchantRuleListOut>('/rules');
}

/** POST /api/rules — 新建。 */
export function createRule(body: MerchantRuleCreate): Promise<MerchantRuleOut> {
  return apiFetch<MerchantRuleOut>('/rules', { method: 'POST', body });
}

/** PATCH /api/rules/{id} — 改 pattern / match_kind / category_id / priority。 */
export function updateRule(
  id: number,
  body: MerchantRuleUpdate,
): Promise<MerchantRuleOut> {
  return apiFetch<MerchantRuleOut>(`/rules/${id}`, { method: 'PATCH', body });
}

/** DELETE /api/rules/{id} — 后端处理被引用情况,前端只关心 204。 */
export function deleteRule(id: number): Promise<void> {
  return apiFetch<void>(`/rules/${id}`, { method: 'DELETE' });
}

File: backend/app/api/rules.py
"""MerchantRules API — spec § 9.1 + § 7。"""
from fastapi import APIRouter, HTTPException, status
from sqlalchemy import select

from app.api.deps import CurrentUserDep, DbDep
from app.models import Category, MerchantRule
from app.schemas import MerchantRuleCreate, MerchantRuleOut, MerchantRuleUpdate
from pydantic import BaseModel


class RuleListOut(BaseModel):
    items: list[MerchantRuleOut]
    total: int


router = APIRouter(prefix="/rules", tags=["rules"])


def _validate_category_or_marker(db, user_id: int, category_id: int | None) -> None:
    if category_id is None:
        return  # marker rule(spec § 7.1)
    cat = db.execute(
        select(Category).where(Category.id == category_id, Category.user_id == user_id)
    ).scalar_one_or_none()
    if cat is None:
        raise HTTPException(status.HTTP_404_NOT_FOUND, "category not found")


@router.get("", response_model=RuleListOut)
def list_rules(user: CurrentUserDep, db: DbDep) -> RuleListOut:
    items = db.execute(
        select(MerchantRule).where(MerchantRule.user_id == user.id)
        .order_by(MerchantRule.priority.asc(), MerchantRule.id.asc())
    ).scalars().all()
    return RuleListOut(
        items=[MerchantRuleOut.model_validate(r) for r in items], total=len(items),
    )


@router.post("", response_model=MerchantRuleOut, status_code=201)
def create_rule(
    body: MerchantRuleCreate, user: CurrentUserDep, db: DbDep,
) -> MerchantRuleOut:
    _validate_category_or_marker(db, user.id, body.category_id)
    rule = MerchantRule(
        user_id=user.id, pattern=body.pattern, match_kind=body.match_kind,
        category_id=body.category_id, priority=body.priority,
    )
    db.add(rule); db.flush()
    return MerchantRuleOut.model_validate(rule)


def _get_rule_or_404(db, user_id: int, rule_id: int) -> MerchantRule:
    r = db.execute(
        select(MerchantRule).where(
            MerchantRule.id == rule_id, MerchantRule.user_id == user_id,
        )
    ).scalar_one_or_none()
    if r is None:
        raise HTTPException(status.HTTP_404_NOT_FOUND, "rule not found")
    return r


@router.patch("/{rule_id}", response_model=MerchantRuleOut)
def update_rule(
    rule_id: int, body: MerchantRuleUpdate, user: CurrentUserDep, db: DbDep,
) -> MerchantRuleOut:
    rule = _get_rule_or_404(db, user.id, rule_id)
    data = body.model_dump(exclude_unset=True)
    if "category_id" in data:
        _validate_category_or_marker(db, user.id, data["category_id"])
    for field, val in data.items():
        setattr(rule, field, val)
    db.flush()
    return MerchantRuleOut.model_validate(rule)


@router.delete("/{rule_id}", status_code=204)
def delete_rule(rule_id: int, user: CurrentUserDep, db: DbDep) -> None:
    rule = _get_rule_or_404(db, user.id, rule_id)
    db.delete(rule); db.flush()
    return None
